### PR TITLE
Add simple assert to detect fake tensors on modules

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89723

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: test/functorch/test_aotdispatch.py [hunk 1]
```diff
@@ -1542,6 +1542,39 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
...
+        real_y = torch.randn(4)
+        fake_mode = torch._subclasses.fake_tensor.FakeTensorMode()
+        fake_y = fake_mode.from_tensor(real_y)
+
+        tracer = torch.fx.Tracer()
+        tracer.record_stack_traces = True
+        graph = tracer.trace(MockModule(fake_y))
+        mod_fake = torch.fx.GraphModule(tracer.root, graph)
+
+        with self.assertRaises(AssertionError):
```
Comment (ezyang) [inline test/functorch/test_aotdispatch.py:1569]: regex the message? Or assertRaisesExpected even better

File: test/functorch/test_aotdispatch.py [hunk 2]
```diff
@@ -1542,6 +1542,39 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
...
+
+        real_x = torch.randn(4)
+        real_y = torch.randn(4)
+        fake_mode = torch._subclasses.fake_tensor.FakeTensorMode()
+        fake_y = fake_mode.from_tensor(real_y)
+
+        tracer = torch.fx.Tracer()
+        tracer.record_stack_traces = True
+        graph = tracer.trace(MockModule(fake_y))
+        mod_fake = torch.fx.GraphModule(tracer.root, graph)
```
Comment (ezyang) [inline test/functorch/test_aotdispatch.py:1567]: I don't understand the test here. What exactly is erroneous about this code?

Some things I might have expected to see here:

1. Fakeifying a module and then passing it to `aot_module_simplified` (should be covered here)
2. A dynamo backend that returns a GraphModule with some fake buffers on it (doesn't look like it's covered here)
3. A bug in tracing that results in a fake tensor constant showing up in the traced joint graph (doesn't look like it's covered here). Probably easiest way to synthesize the bug is to pass a function that directly accesses a fake tensor from inside the tracing context.

File: functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -1742,6 +1742,8 @@ def aot_module_simplified(
     """
     #########################################################
 
+    torch._dynamo.utils.assert_no_fake_params_or_buffers(mod)
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:1745]: A little comment explaining why it's invalid to pass fakeified modules to aot_module_simplified would be nice!

**Review by ezyang — APPROVED** [review]

File: test/functorch/test_aotdispatch.py (hunk 1 above)
Comment (voznesenskym) [reply test/functorch/test_aotdispatch.py:1569]: sure

File: test/functorch/test_aotdispatch.py (hunk 2 above)
Comment (voznesenskym) [reply test/functorch/test_aotdispatch.py:1567]: (3) is not covered in this PR - this is purely for the lowering component - we need a different PR for the joint graph because I think thats a little trickier. We could detect it on the bw_module, but I think it might be better to do it at a different level, where we raise up the buffers and params into inputs.

File: torch/_dynamo/optimizations/distributed.py [hunk 4]
```diff
@@ -242,7 +243,7 @@ def forward(self, *args):
                             unwrap_singleton_tuple = True
                             sn.args = (sn.args,)
                 submod.recompile()
-
+                assert_no_fake_params_or_buffers(submod)
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:246]: Agreed, done.

File: test/functorch/test_aotdispatch.py (hunk 2 above)
Comment (voznesenskym) [reply test/functorch/test_aotdispatch.py:1567]: Oh and 

>  What exactly is erroneous about this code?

self.y is fake here, which is illegal.

Comment (voznesenskym) [reply test/functorch/test_aotdispatch.py:1567]: > A dynamo backend that returns a GraphModule with some fake buffers on it (doesn't look like it's covered here)

Dynamo backends return functions tho, so while we could do this, the hook here would be clumsy.

Comment (ezyang) [reply test/functorch/test_aotdispatch.py:1567]: I don't think this level exists? If I add a constant into graphmodule from fx tracing, it always stays a constant, it never gets input-ified. I think we have to do some AOTAutograd specific test here; or maybe `make_fx`, but generic usage of `make_fx` COULD validly put fake tensors onto the returned GraphModule so you'd have to think about when exactly this invariant holds.

File: functorch/_src/aot_autograd.py (hunk 3 above)
Comment (voznesenskym) [reply functorch/_src/aot_autograd.py:1764]: Done

File: test/functorch/test_aotdispatch.py (hunk 2 above)
Comment (voznesenskym) [reply test/functorch/test_aotdispatch.py:1567]: Hmm, maybe I misunderstood you - are you suggesting we add this invariant to outputs? If so, I think this would be a new invariant. I think checking the inputs on the graph module covers the cases we ran into well enough, and I would rather keep this change conservative.

Comment (ezyang) [reply test/functorch/test_aotdispatch.py:1567]: > self.y is fake here, which is illegal.

Oh, I missed this. Would be a good idea to call it out in the test.

> Hmm, maybe I misunderstood you - are you suggesting we add this invariant to outputs? If so, I think this would be a new invariant. I think checking the inputs on the graph module covers the cases we ran into well enough, and I would rather keep this change conservative.

What I am saying is that when I get a GraphModule from make_fx, in the context of AOTAutograd this GraphModule should NEVER have fake tensors on it. This is NOT true for make_fx calls in general, but for Dynamo usage of AOTAutograd we do maintain this invariant.

File: torch/_dynamo/optimizations/distributed.py [hunk 5]
```diff
@@ -242,7 +242,6 @@ def forward(self, *args):
                             unwrap_singleton_tuple = True
                             sn.args = (sn.args,)
                 submod.recompile()
-
```
Comment (ezyang) [inline torch/_dynamo/optimizations/distributed.py:245]: undo this change

File: functorch/_src/aot_autograd.py [hunk 6]
```diff
@@ -1742,6 +1749,20 @@ def aot_module_simplified(
     """
     #########################################################
 
+    # Redudant with dynamo, but worth having in case this gets invoked elsewhere.
+
+    # Fake Modules and AOTAutograd
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:1754]: The prevailing convention is `Note [Fake Modules and AOTAutograd]`, this is mostly to clue future code writers in that the title matters and if they rename it they have to change the reference sites too

File: functorch/_src/aot_autograd.py [hunk 7]
```diff
@@ -1742,6 +1749,20 @@ def aot_module_simplified(
...
+    #
+    # Passing a module with *any* fake tensors on it is invalid, as it will
+    # fake-tensor poison the runtime component of aot_autograd.
+    # Compilation will be done with fake tensors, and runtime will be invoked with real tensors.
+    # The module passed in here is utilized again as input for compiling the
+    # fx graph, from the joint forward backward function passed to make_fx.
+    # If a fake tensor got into the fx graph, it will get into the fw_module or bw_module
+    # and we will fail due to mixed real/fake tensors in runtime, or, produce a fake tensor output which will
+    # poison the next subsequent call.
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:1763]: I think we can make this comment more clear. How about this:

> A simple heuristic for when to use fake versus real tensors is that fake tensors are for compile time (when we don't want to actually run the compute, but we do want to know about metadata), and real tensors are for runtime (when we actually want to do the compute.) However, in AOTAutograd, modules are the exception: we always pass AOTAutograd modules with real tensors. This is because AOTAutograd will produce a compiled function which needs to directly access any parameters the compiled function may need, but these parameters will NOT be passed in by the caller (aka Dynamo). So at compile time, the compiled function we produce must close over any parameters, and those parameters must be real parameters, and we cannot do this unless at compile time we get a module with real tensors.

> Even if Dynamo did pass all parameters explicitly at runtime, which would eliminate the need to close over the parameters, it would still be profitable to pass real tensor parameters to the compiler at compile time, because some compilation strategies like CUDA graphs want to burn in the pointer addresses where the parameter data live, and of course we can't do that unless we give the backend a real tensor.

File: test/functorch/test_aotdispatch.py [hunk 8]
```diff
@@ -1542,6 +1542,69 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
...
+                z = z.relu()
+                return (z, )
+
+
+        real_x = torch.randn(4)
+        real_y = torch.randn(4)
+
+        tracer = torch.fx.Tracer()
+        tracer.record_stack_traces = True
+        graph = tracer.trace(MockModule(real_y))
```
Comment (ezyang) [inline test/functorch/test_aotdispatch.py:1597]: Pretty sure you don't need this, regular modules will deepcopy to fake just fine too

Comment (voznesenskym) [reply test/functorch/test_aotdispatch.py:1597]: sorry, what do you mean?

Comment (ezyang) [reply test/functorch/test_aotdispatch.py:1589]: You're running FX symbolic trace in this test but it's totally unnecessary (and in the conventional dynamo+aot autograd flow you never use a standard tracer.trace)

Comment (voznesenskym) [reply test/functorch/test_aotdispatch.py]: Fixed one test, left this one as is on purpose, added comment.
